sn_cpu.slurm

```diff
@@ -0,0 +1,26 @@
+#!/bin/bash
+#
+#SBATCH --account=watertemp
+#SBATCH --partition=cpu
+#SBATCH --job-name=snakemake
+#SBATCH --time=1-23:59:59
+#SBATCH --ntasks=1
+#SBATCH --cpus-per-task=16
+#SBATCH --output=log/sbatch_all_%a_%A.out
+#SBATCH --error=log/sbatch_all_%a_%A.out
+
+## Initialize work environment
+# conda activate ltls
+# which python
+module load analytics cuda11.3/toolkit/11.3.0
+
+## Run the main job
+
+run_id=test1
+model_id=cpu_a
+log_dir=log/${run_id}_${model_id}
+mkdir -p ${log_dir}
+# From Jeff's blogpost: https://github.com/jsadler2/wdfn-blog/blob/ca63d4dc4bf6aeb6962a58953057a41c9db1ad8a/content/snakemake-ml-experiments.md
+# Also see snakemake profiles: https://snakemake.readthedocs.io/en/stable/executing/cli.html#profiles
+snakemake --cluster "srun -A watertemp -t 1-23:59:59 -p cpu -N 1 -n 1 -c 16 --job-name=${run_id}_${model_id} -e ${log_dir}/slurm-%j.out -o ${log_dir}/slurm-%j.out" --printshellcmds --keep-going --cores all --jobs 1 --rerun-incomplete 3_train/out/mntoha/${run_id}/${model_id}_weights.pt 2>&1 | tee ${log_dir}/run.out
```

Comment:
I've not launched parallel tasks with slurm before, so I'm not sure that I can review this bit - you might want to run it by Jake or Jeff or Jeremy?

If it's running as expected as is, maybe this doesn't even need to be reviewed?

_______________________________

Here are my thoughts (keep that ^ caveat in mind):

When you launch this are you running `sh setup.sh` followed by `sbatch sn_cpu.slurm` (if running on the cpu partition), or `sbatch sn_gpu.slurm` (if running on the gpu portion)?

Here, you've set up a sbatch script in a `.slurm` file. The sbatch script requests an allocation, specifying 8 options (`-A`, `-p`, `-J`, `-t`, `-n`, `-c`, `-o`, and `-e`). Once it has that allocation, it submits a `Snakemake` `--cluster` command that includes a `srun` command that includes a _second_ allocation request, this time specifying 9 options (adding in the `-N` option to specify the number of nodes).

I'm a bit unsure about what's happening here with the double allocation request. It seems like the use of `srun` is right, per [the docs](https://hpcportal.cr.usgs.gov/training/courses/Intro_to_HPC/10_slurm.html#priority-reservations):
> **Execution with srun**
`srun [OPTIONS(0)...] [ : [OPTIONS(N)...]] executable(0) [args(0)...]`
> * Runs a parallel job on cluster managed by Slurm. If necessary, srun will first create a resource allocation in which to run the parallel job.
> * Within an allocation, srun is used to determine what runs on any given subset of resources available
> * Multiple srun commands can be used sequentially or concurrently within a job
> * Failing to use srun within a job allocation will result in the executables being run only on a single node, with an undetermined set of resources. This leads to resource contention and generally bad behavior

...but I'm confused about what that allocation request within the `srun` command is doing. My initial thinking was, given that the `srun` allocation request is identical to the `sbatch` allocation request, is it needed here? But would guess there's a reason you included a second allocation request, and suspect I'm just missing some nuance of slurm. When you run this, what job names, etc. do you see in `squeue`? Are they named `snakemake`, or `test1_cpu_a`? Or is there a single `snakemake` job and then many `test1_cpu_a` jobs?  Also I see that you don't specify the number of nodes in your `sbatch` request, but you do in  your `srun` request. Since the default is 1 task per node and you're only requesting `-n 1` does it not matter?

_To help me think this through, I checked out these other sbatch uses:_
  * In Jeff's blog post, to run tasks in parallel, he submitted a [sbatch command](https://github.com/jsadler2/wdfn-blog/blob/snakemake-dl-experiments/content/snakemake-ml-experiments.md#executing-snakemake-on-many-cores) _within_ a `Snakemake` command

  * For `lake-temperature-neural-networks`, to run tasks in serial, we ran an R script in a [`slurm.sbatch` file](https://code.usgs.gov/hcorson-dosch/lake-temperature-neural-networks/-/blob/master/2_model/src/slurm.sbatch) by [calling `sbatch 2_model/src/slurm.sbatch` in the terminal](https://code.usgs.gov/hcorson-dosch/lake-temperature-neural-networks/-/blob/master/2_model/src/model_tasks.R#L139-142).